src/txrebroadcast.cpp

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <chainparams.h>
+#include <consensus/consensus.h>
+#include <logging.h>
+#include <miner.h>
+#include <script/script.h>
+#include <txrebroadcast.h>
+#include <util/time.h>
+#include <validation.h>
+
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index_container.hpp>
+
+/** We rebroadcast 3/4 of max block weight to reduce noise due to circumstances
+ *  such as miners mining priority transactions. */
+static constexpr unsigned int MAX_REBROADCAST_WEIGHT{3 * MAX_BLOCK_WEIGHT / 4};
+
+/** Default minimum age for a transaction to be rebroadcast */
+static constexpr std::chrono::minutes REBROADCAST_MIN_TX_AGE{30min};
+
+/** Maximum number of times we will rebroadcast a transaction */
+static constexpr int MAX_REBROADCAST_COUNT{6};
+
+/** Minimum amount of time between returning the same transaction for
+ * rebroadcast */
+static constexpr std::chrono::hours MIN_REATTEMPT_INTERVAL{4h};
+
+/** The maximum number of entries permitted in m_attempt_tracker */
+static constexpr int MAX_ENTRIES{500};
+
+/** The maximum age of an entry ~3 months */
+static constexpr std::chrono::hours MAX_ENTRY_AGE{24h * 30 * 3};
```

Comment:
I don't understand this comment, let me explain some of the mechanisms incase that helps. otherwise, please help me understand what you are trying to communicate.

> I don't think Bitcoin applications care about confirmation of transactions broadcast 3 months ago.

1. the attempt tracker records the timestamp of when the mempool transaction was last selected for rebroadcast, not when it was initially broadcast / ATMPed (see https://github.com/bitcoin/bitcoin/pull/21061/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=.py&file-filters%5B%5D=.sh#diff-7dff50848db96bdb8edffc4d21daeca6d9050ec0e67d96072780ea5751e7df06R110 and https://github.com/bitcoin/bitcoin/pull/21061/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=.py&file-filters%5B%5D=.sh#diff-7dff50848db96bdb8edffc4d21daeca6d9050ec0e67d96072780ea5751e7df06R125).
2. `MAX_ENTRY_AGE` is only called from `TrimMaxRebroadcast`. It is essentially used to say "if a transaction hasn't been selected for rebroadcast in the past 3 months, remove it from the attempt tracker"

> Let's say in optimistic scenarios, a rebroadcast entry has a compelling feerate enough to be selected by GetRebroadcastTransactions after each MIN_REATTEMPT_INTERVAL, after 6 * 4, it's matured enough to be legitimately expired ?

can you clarify what you mean by "legitimately expired"? if you mean that we should disable automatic rebroadcast so the txn can (probably) expire from the (majority of) network mempools, then yes I agree. which is why we put it on the tracker. but then, what is the question? 

>  select MEMPOOL_EXPIRY as a maximum age.

if we stored transactions on the attempt tracker for the same duration as `MEMPOOL_EXPIRY` time, that would leave room for the issue that the attempt tracker is meant to address. for more explanation on that, please see https://github.com/amitiuttarwar/bitcoin-notes/blob/main/rebroadcast-history.md#faq section "What happens if a transaction can / will never be mined?". expiry would be one of the main ways to get the ping-ponging behavior for never-going-to-be-mined transactions (eg. due to a policy upgrade on the network).

Reply:
IIRC, by "legitimately expired" I meant a transaction of which `m_count` has reached `MAX_REBROADCAST_COUNT`. As we already rebroadcasted this transaction the maximum number of times permitted, we skip its rebroadcasting forever, but without necessarily removing it from `m_attempt_tracker` ? I think such transaction has exhausted its rebroadcast "credit" and it's mature enough to be "legitimately" expired or pruned. Though this deletion criteria isn't currently considered in `TrimMaxRebroadcast()` ? 

I think this deletion criteria would be more efficient to expire transactions from the majority of network mempools rather than MAX_ENTRY_AGE, especially with a 3 months delay.

> if we stored transactions on the attempt tracker for the same duration as MEMPOOL_EXPIRY time, that would leave room for the issue that the attempt tracker is meant to address. for more explanation on that, please see https://github.com/amitiuttarwar/bitcoin-notes/blob/main/rebroadcast-history.md#faq section "What happens if a transaction can / will never be mined?". expiry would be one of the main ways to get the ping-ponging behavior for never-going-to-be-mined transactions (eg. due to a policy upgrade on the network)

IIUC, the problem we're trying to solve is "How to avoid never-going-to-be-mined transactions to forever ping-pong across the network ?". To achieve this aim, I believe we should expire a rebroadcast entry
as soon as it has exhausted its rebroadcast count, a hint that either feerate is definitively aloof or severer circumstances as described in documentation. But selecting a MAX_ENTRY_AGE superior to 
MEMPOOL_EXPIRY moves in the opposite direction. If rebroadcast entry _average_ rebroadcast interval  is superior to 56 hours, 56 * MAX_REBROADCAST_COUNT== MEMPOOL_EXPIRY and node rebroadcast
guarantees attempts beyond the mempool expiration limit ?

Or is the behavior I'm describing not possible ? I'm happy to illustrate with a test to assert our intuitions :)

Reply:
it seems like the majority of this comment is based on a misunderstanding of how `m_attempt_tracker` works. 

the update seems like its getting on the right track- the main purpose of tracking rebroadcast entries is to allow enforcing a maximum number of rebroadcast attempts. 

> IIUC, the problem we're trying to solve is "How to avoid never-going-to-be-mined transactions to forever ping-pong across the network ?".

yes, exactly.

but it still seems like there is remaining confusion. `MAX_ENTRY_AGE` indicates the amount of time has passed since tx A has been rebroadcast to peers. if its been 3 months since we last tried to rebroadcast tx A, we remove the attempts from the tracker. if there is no entry for tx A on the tracker, we don't have record of the fact that we tried to rebroadcast it, so if tx A is selected by the block assembler, we would rebroadcast it again. 

I don't see how `MAX_ENTRY_AGE` would viably be used for node fingerprinting when you combine with the other factors such as `filterInventoryKnown`.